sql/2024/08/performance-lab-plugins-adoption.sql

```diff
@@ -0,0 +1,108 @@
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# See https://github.com/GoogleChromeLabs/wpp-research/pull/146
+
+DECLARE
+  DATE_TO_QUERY DATE DEFAULT '2024-07-01';
+
+CREATE TEMPORARY FUNCTION HAS_TECHNOLOGY(technologies ARRAY<STRUCT<technology STRING, categories ARRAY<STRING>, info ARRAY<STRING>>>, technologyToFind STRING) RETURNS BOOL AS (
+  EXISTS(
+    SELECT
+      *
+    FROM
+      UNNEST(technologies) AS technology
+    WHERE technology.technology = technologyToFind
+  )
+);
+
+CREATE TEMPORARY FUNCTION GET_GENERATOR_CONTENTS(custom_metrics STRING) RETURNS ARRAY<STRING> AS (
+  ARRAY(
+    SELECT
+      JSON_EXTRACT_SCALAR(metaNode, '$.content') AS content
+    FROM
+      UNNEST(JSON_EXTRACT_ARRAY(JSON_EXTRACT(custom_metrics, '$.almanac.meta-nodes.nodes'))) AS metaNode
+    WHERE
+      JSON_EXTRACT_SCALAR(metaNode, '$.name') = 'generator'
+  )
+);
+
+CREATE TEMPORARY FUNCTION HAS_PERFORMANCE_LAB(contents ARRAY<STRING>) RETURNS BOOL AS (
+  EXISTS(
+    SELECT
+      *
+    FROM
+      UNNEST(contents) AS content
+    WHERE
+      STARTS_WITH(content, 'performance-lab ')
+  )
+);
+
+CREATE TEMPORARY FUNCTION EXTRACT_PL_PLUGINS_FROM_GENERATOR_CONTENTS(contents ARRAY<STRING>) RETURNS ARRAY<STRING> AS (
+  ARRAY(
+    SELECT
+      REGEXP_EXTRACT(content, '^([a-z0-9-]+) ') AS slug
+    FROM
+      UNNEST(contents) as content
+    WHERE
+      REGEXP_EXTRACT(content, '^([a-z0-9-]+) ') IN UNNEST([
+        'auto-sizes',
+        'dominant-color-images',
+        'embed-optimizer',
+        'image-prioritizer',
+        'optimization-detective',
+        'performant-translations', # Developed in separate repository.
+        'speculation-rules',
+        'web-worker-offloading',
+        'webp-uploads'
+      ])
+  )
+);
+
+WITH urlsWithPerformanceLabPlugins AS (
+  SELECT
+    client,
+    page,
+    # COUNT(DISTINCT page) OVER (PARTITION BY client) AS total,
+    HAS_PERFORMANCE_LAB(GET_GENERATOR_CONTENTS(custom_metrics)) AS has_pl,
+    EXTRACT_PL_PLUGINS_FROM_GENERATOR_CONTENTS(GET_GENERATOR_CONTENTS(custom_metrics)) AS plugins
+  FROM
+    `httparchive.all.pages`
+  WHERE
+    date = DATE_TO_QUERY
+    AND HAS_TECHNOLOGY(technologies, 'WordPress')
+    AND is_root_page = TRUE
+    AND ARRAY_LENGTH(EXTRACT_PL_PLUGINS_FROM_GENERATOR_CONTENTS(GET_GENERATOR_CONTENTS(custom_metrics))) > 0
+)
+
+SELECT
+  client,
+  plugin,
+  COUNT(DISTINCT page) AS urls,
+  COUNT(DISTINCT IF(has_pl = TRUE, page, NULL)) AS urls_with_pl,
+  COUNT(DISTINCT IF(has_pl = FALSE, page, NULL)) AS urls_without_pl
+  # ANY_VALUE(total) AS total,
+  # COUNT(DISTINCT page) / ANY_VALUE(total) AS pct_total
```

Comment:
It would be nice to include percentages in the report. Could this be updated to show the percentage that includes PL, i.e. `urls_with_pl / urls`?

Reply:
@joemcgill I actually missed to remove those commented-out parts, they were from another query I used as the foundation for this one, which was querying module adoption relative to overall PL (since back then the modules of course couldn't be used without PL). But now that part's no longer relevant so this code wouldn't even work if un-commented. I'll remove it.

I can add something simple though to show for each plugin the % of "URLs with PL active" / "Overall URLs", which of course now can easily be calculated already, but I see how having it in the result would be better.